frontend/device/kobo/powerd.lua

```diff
@@ -300,35 +305,81 @@ function KoboPowerD:isChargedHW()
+    end
+end
+
+-- This ramp down goes faster on high intensity and slower at the end.
+-- The user will notice a linear brightness change.
+-- The whole function gets called at max log(100)/log(0.75) = 17 times,
+-- which will lead to 0.025*17 + 0.5 = 0.925s ramp down time (non blocking), can be aborted.
+function BasePowerD:turnOffFrontlightRamp(curr_ramp_intensity, end_intensity)
+    curr_ramp_intensity = math.floor(math.max(curr_ramp_intensity * .75, 0))
+
+    if curr_ramp_intensity > end_intensity then
+        self:_setIntensityHW(math.floor(curr_ramp_intensity))
+        UIManager:scheduleIn(0.025, self.turnOffFrontlightRamp, self, curr_ramp_intensity, end_intensity)
+    else
+        -- On some devices (Sage) setting intensity to zero happens immediately,
+        -- which will lead to a jump and not a ramp. So we postpone the last change by 0.5s to get it smooth
+        UIManager:scheduleIn(self.device.frontlight_settings.wait_before_turn_off_s or 0.0, self._postponedSetIntensityHW, self, end_intensity)
+        -- no reschedule here, as we are done
+    end
+end
+
 function KoboPowerD:turnOffFrontlightHW()
     if not self:isFrontlightOnHW() then
         return
     end
-    ffiUtil.runInSubProcess(function()
-        for i = 1, 5 do
-            -- NOTE: Do *not* switch to (self.fl_intensity * (1/5) * i) here, it may lead to rounding errors,
-            --       which is problematic paired w/ math.floor because it doesn't round towards zero,
-            --       which means we may end up passing -1 to setIntensityHW, which will fail,
-            --       because we're bypassing the clamping usually done by setIntensity...
-            self:setIntensityHW(math.floor(self.fl_intensity - (self.fl_intensity / 5 * i)))
-            --- @note: Newer devices appear to block slightly longer on FL ioctls/sysfs, so only sleep on older devices,
-            ---        otherwise we get a jump and not a ramp ;).
-            if not self.device:hasNaturalLight() then
-                if (i < 5) then
-                    ffiUtil.usleep(35 * 1000)
-                end
+
+    if UIManager then
+        -- Do nothing, if the ramp is allready running
+        if not self.fl_ramp_down_running then
+            self:_stopFrontlightRamp()
+            self:turnOffFrontlightRamp(self.fl_intensity, self.fl_min)
+            self.fl_ramp_down_running = true
+
+            -- NOTE: This is essentially what setIntensityHW does, except we don't actually touch the FL,
+            --       we only sync the state of the main process with the final state of what we're doing in the forks.
+            -- And update hw_intensity in our actual process ;).
+            self.hw_intensity = self.fl_min
+            -- NOTE: And don't forget to update sysfs_light, too, as a real setIntensityHW would via setBrightness
+            if self.fl then
+                self.fl.current_brightness = self.fl_min
             end
+            self:_decideFrontlightState()
```

Comment:
This shouldn't be necessary anymore, since everything happens in the main thread now ;).

(Ditto for the other ramp).

Reply:
Okay, yeah, that makes sense; I'd possibly simply rejig the comments a bit then ;).